File: 프로그래머스/2/250136. ［PCCP 기출문제］ 2번 ／ 석유 시추/［PCCP 기출문제］ 2번 ／ 석유 시추.py
from collections import deque

dy = [-1, 1, 0, 0]
dx = [0, 0, -1, 1]
    
def solution(land):
    row, col = len(land), len(land[0])
    visited = [[0] * col for _ in range(row)]
    song = [0] * col
    
    for i in range(row):
        for j in range(col):
            if not visited[i][j] and land[i][j] == 1:
                bfs(i, j, visited, row, col, land, song)
                
    return max(song)
    

def bfs(y, x, visited, row, col, land, song):
    queue = deque([(y, x)])
    visited[y][x] = 1
    amount = 0
    x_visited = set([x])

    while queue:
        new_y, new_x = queue.popleft()
        amount += 1

        for i in range(4):
            move_y = new_y + dy[i]
            move_x = new_x + dx[i]

            if 0 <= move_y < row and 0 <= move_x < col and not visited[move_y][move_x] and land[move_y][move_x] == 1:
                queue.append((move_y, move_x))
                x_visited.add(move_x)
                visited[move_y][move_x] = 1
                
    for i in x_visited:
        song[i] += amount
    

File: 프로그래머스/2/250136. ［PCCP 기출문제］ 2번 ／ 석유 시추/［PCCP 기출문제］ 2번 ／ 석유 시추.java
import java.util.*;

class Solution {
    
    boolean[][] visited; 
    int[] dx = new int[]{1, 0, -1, 0};
    int[] dy = new int[]{0, 1, 0, -1};
    int N, M;
    Map<Integer, Integer> map = new HashMap<>();
    public int solution(int[][] land) {
        int answer = 0;
        N = land.length;
        M = land[0].length;
        visited = new boolean[N][M];
        int index = 1;
        for(int i =0; i<N; i++) {
            for(int j=0; j<M; j++) {
                if (land[i][j] == 0 || visited[i][j]) {
                    continue;
                }
                
                int count = bfs(land, i, j, index);
                map.put(index, count);
                index++;
            }
        }
        
        for(int j=0; j<M; j++) {
            Set<Integer> set = new HashSet<>();
            int sum = 0;
            for(int i=0; i<N; i++) {
                if (land[i][j] == 0) {
                    continue;
                }
                
                set.add(land[i][j]);
            }
            
            for(int key: set) {
                sum += map.get(key);
            }
            
            answer = Math.max(answer, sum);
        }
        
        return answer;
    }
    
    private int bfs(int[][] land, int i, int j, int index) {
        if (visited[i][j]) {
            return 0;
        }  
        
        int sum = 1;
        Deque<int[]> deque = new ArrayDeque<>();
        deque.add(new int[]{i, j});
        visited[i][j] = true;
        land[i][j] = index;
        while(!deque.isEmpty()) {
            int[] temp = deque.pollFirst();
            int currentI = temp[0];
            int currentJ = temp[1];
            
            for(int k=0; k<4; k++) {
                int nextI = currentI + dx[k];
                int nextJ = currentJ + dy[k];
                
                if(nextI < 0 || nextI > N-1 || nextJ < 0 || nextJ > M-1) {
                    continue;
                }
                
                if(visited[nextI][nextJ]) {
                    continue;
                }
                
                if(land[nextI][nextJ] == 0) {
                    continue;
                }
                
                deque.add(new int[]{nextI, nextJ});
                visited[nextI][nextJ] = true;
                land[nextI][nextJ] = index;
                sum++;
            }
        }
    
        return sum;
    }
}